### ISSUE TITLE
Stopping Nemesis on Exception

### ISSUE THREAD

**Issue by korry8911** [jepsen-io/jepsen#321]
I have a Nemesis that will produces and exception occasionally. I want to have the test abort if the Nemesis produces this specific exception but right now, all my exception are getting caught and the test continues. If the Nemesis throws an exception and does not abort the test, the final results may be misleading in my scenario. I have looked through the code but could find no obvious way to stop the test and it looks like every exception is being caught by try-catch throwable in run!.  Is there a simple way to accomplish this without modifying the Jepsen code itself? Thanks, Korry

**Comment by aphyr** [comment]
It sounds like you'd like to have the generator stop emitting events if the nemesis throws. You can do this by having a mutable atom that you share between the nemesis and the generator, which is `true` as long as you'd like the test to run. The generator could check that atom's state, and if it's true, go on to call the rest of the generator for your test. The nemesis, meanwhile, could catch that exception, and in the catch block, reset the atom to false, causing the test to end.

**Comment by korry8911** [comment]
Thanks, yeah I was thinking that might be the way to do it. But I was wondering, if the nemesis resets this atom to stop the generator, the test will end and the analysis would happen. My goal is to have the test end, and always fail without doing the analysis. I'll explore this some more. Thanks for the quick response. 

Side question: any plan to create a sequential consistency and causal consistency checker to supplement the linearizability checker?

**Comment by aphyr** [comment]
Oh! Uhhm... So generally the idea with Jepsen is that you're trying to measure whether invariants hold under nemesis operations! I'm sure you've got a good reason for doing this--just want to let you know this is a bit of a hack. I thiiink you should be able to throw in the generator, and that ought to cause the test to abort.For sequential and causal checkers, it'd be nice, but no plans immediately. I'm not sure how to do it yet!

**Comment by korry8911** [comment]
Just wanted to give an update in case anyone else has this issue.

So the solution we found was to add a control atom (:continue or :abort) to the test map and wrap our client generator in a function that will continue producing ops as long as the control atom equals :continue. We also had to override invoke-compat! to catch exceptions thrown by the nemesis, set the control atom to :abort, and rethrown the exception. This causes the client generators to stop producing ops and the entire test to exit. Then we added a sanity checker to check if the control atom has been set to :abort, in which case it will override the Jepsen analysis and automatically fail the test.
